tests/sccache_cargo.rs

```diff
@@ -64,74 +112,123 @@ fn test_rust_cargo_cmd(cmd: &str, extra_envs: &[(&str, &std::ffi::OsStr)]) {
+        "check",
+        SccacheTest::new(Some(&[("RUSTFLAGS", OsString::from("-Zprofile"))]))?,
+    )?;
+    #[cfg(feature = "unstable")]
+    test_rust_cargo_cmd(
+        "build",
+        SccacheTest::new(Some(&[("RUSTFLAGS", OsString::from("-Zprofile"))]))?,
+    )?;
+
+    Ok(())
+}
+
+fn test_rust_cargo_cmd(cmd: &str, test_info: SccacheTest) -> Result<()> {
     // `cargo clean` first, just to be sure there's no leftover build objects.
-    let mut envs = vec![
-        ("RUSTC_WRAPPER", sccache.as_ref()),
-        ("CARGO_TARGET_DIR", cargo_dir.as_ref()),
-        // Explicitly disable incremental compilation because sccache is unable
-        // to cache it at the time of writing.
-        ("CARGO_INCREMENTAL", OsStr::new("0")),
-    ];
-    envs.extend_from_slice(extra_envs);
-    Command::new(&cargo)
+    Command::new(CARGO.as_os_str())
         .args(&["clean"])
-        .envs(envs.iter().copied())
-        .current_dir(&crate_dir)
+        .envs(test_info.env.iter().cloned())
+        .current_dir(CRATE_DIR.as_os_str())
         .assert()
-        .success();
+        .try_success()?;
     // Now build the crate with cargo.
-    Command::new(&cargo)
+    Command::new(CARGO.as_os_str())
         .args(&[cmd, "--color=never"])
-        .envs(envs.iter().copied())
-        .current_dir(&crate_dir)
+        .envs(test_info.env.iter().cloned())
+        .current_dir(CRATE_DIR.as_os_str())
         .assert()
-        .stderr(predicates::str::contains("\x1b[").from_utf8().not())
-        .success();
+        .try_stderr(predicates::str::contains("\x1b[").from_utf8().not())?
+        .try_success()?;
     // Clean it so we can build it again.
-    Command::new(&cargo)
+    Command::new(CARGO.as_os_str())
         .args(&["clean"])
-        .envs(envs.iter().copied())
-        .current_dir(&crate_dir)
+        .envs(test_info.env.iter().cloned())
+        .current_dir(CRATE_DIR.as_os_str())
         .assert()
-        .success();
-    Command::new(&cargo)
+        .try_success()?;
+    Command::new(CARGO.as_os_str())
         .args(&[cmd, "--color=always"])
-        .envs(envs.iter().copied())
-        .current_dir(&crate_dir)
+        .envs(test_info.env.iter().cloned())
+        .current_dir(CRATE_DIR.as_os_str())
         .assert()
-        .stderr(predicates::str::contains("\x1b[").from_utf8())
-        .success();
+        .try_stderr(predicates::str::contains("\x1b[").from_utf8())?
+        .try_success()?;
+
     // Now get the stats and ensure that we had a cache hit for the second build.
     // The test crate has one dependency (itoa) so there are two separate
     // compilations.
     trace!("sccache --show-stats");
-    sccache_command()
+    Command::new(SCCACHE_BIN.as_os_str())
+        .args(&["--show-stats", "--stats-format=json"])
+        .assert()
+        .try_stdout(predicates::str::contains(r#""cache_hits":{"counts":{"Rust":2}}"#).from_utf8())?
+        .try_success()?;
+
+    drop(test_info);
```

Comment:
Isn't `test_info` dropped automatically at the end of this function?
If we need to explicitly stop the server to make it shut down between tests, perhaps we should do so explicitly without leaning on `drop()`?

Reply:
> Isn't test_info dropped automatically at the end of this function?

Yes. But I was/am not sure if it might even be dropped _before_ the end of the function. I was thinking that by explicitly dropping it at the end, I could prevent it from being dropped after its last usage, since the server still has to be alive for the "get stats" step. I guess I probably should make the "show stats" invocation an associated function of `test_info`, that way the server would be guaranteed to still be alive.